### PR TITLE
[FIX] warnings on gcc9

### PR DESCRIPTION
**Pull request by eseiler**
GCC9 introduces new rules when implicitly defaulted copy constructor and copy assignment are valid.

If there is a user-provided copy constructor, the copy assignment has to be provided too, and vice versa.
Additionally, a user-provided destructor requires both copy constructor and assignment to be defined.

Violating this rule results in a `deprecated-copy` warning (e.g. when using `SeqFileIn`):
```
seqan/include/seqan/stream/guess_format.h:159:14: warning: implicitly-declared [COPYCTOR] is deprecated [-Wdeprecated-copy]
  159 |     TLowcase lowcaseFileName(fileName);

seqan/include/seqan/sequence/segment_prefix.h:117:5: note: because 'seqan::Segment<const char*, seqan::PrefixSegment>' has user-provided [COPYASSIGNMENT]
  117 |     operator = (Segment const & source)
```
Where COPYCTOR and COPYASSIGNMENT are the abbreviations of the usually issued full signatures. 

So in this case, we need to add `Segment(Segment const &) = default;`.

### CODE REVIEW

**Comment by rrahn** [comment]
> If you are touching these constructors, we might as well want the rule of five? 

No, if we decide to do that for SeqAn at some point than we would add general support for move awareness. This is not part of this PR yet.

**Comment by rrahn** [comment]
@eseiler can you do this soon? I need to run g++ on macos builds and the warnings need to be fixed. Thanks a lot.

**Comment by eseiler** [comment]
@rrahn should be good now, no warnings on gcc9 and the failing `build_manual` test is your PR, I guess.

File: include/seqan/index/index_qgram.h [hunk 1]
```diff
@@ -275,15 +275,9 @@ class Index<TText_, IndexQGram<TShapeSpec, TSpec> > {
     cargo(other.cargo),
     stepSize(1) {}
 
-    Index(Index const &other):
```
Comment (rrahn) [inline include/seqan/index/index_qgram.h]: do you know why there is another copy constructor that takes the index as non const reference?

File: include/seqan/index/index_qgram_openaddressing.h [hunk 2]
```diff
@@ -107,7 +107,7 @@ namespace seqan
         TShape            shape;        // underlying shape
         TCargo            cargo;        // user-defined cargo
         TBucketMap        bucketMap;    // bucketMap table (used by open-addressing index)
-        TSize            stepSize;    // store every <stepSize>'th q-gram in the index
+        TSize            stepSize{1};   // store every <stepSize>'th q-gram in the index
```
Comment (rrahn) [inline include/seqan/index/index_qgram_openaddressing.h]: why was this necessary? but alpha could be left as is?

File: include/seqan/index/index_qgram.h [hunk 3]
```diff
@@ -247,7 +247,7 @@ class Index<TText_, IndexQGram<TShapeSpec, TSpec> > {
     TShape            shape;        // underlying shape
     TCargo            cargo;        // user-defined cargo
     TBucketMap        bucketMap;    // bucketMap table (used by open-addressing index)
-    TSize            stepSize;    // store every <stepSize>'th q-gram in the index
+    TSize            stepSize{1};   // store every <stepSize>'th q-gram in the index
```
Comment (rrahn) [inline include/seqan/index/index_qgram.h]: why was this necessary?

File: include/seqan/find/find_myers_ukkonen.h [hunk 4]
```diff
@@ -856,7 +856,7 @@ finline void
 _myersPreInit(PatternState_<TNeedle, TSpec> &state, False)
 {
     typedef typename Value<TNeedle>::Type TValue;
-    memset(state.bitMasks, 0, (ValueSize<TValue>::VALUE + 1) * sizeof(state.bitMasks[0]));
+    memset(state.bitMasks, 0, ValueSize<TValue>::VALUE * sizeof(state.bitMasks[0]));
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: It should be `|sigma| + 1`, because other states also set it to this value:

https://github.com/seqan/seqan/blob/30667bc518fa58673d832b60ffb750f57599cb4c/include/seqan/find/find_myers_ukkonen.h#L546-L549

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: If this triggers a warning again, we have to look at it again.

[the remark this answers is not in the document]
Comment (eseiler) [reply include/seqan/find/find_myers_ukkonen.h]: It does. It's accessing 768 in an array of size 256 sometimes. So removing the 1 fixed it, but I didn't have an extensive look at this. Anyway, if it was wrong, I'd hope that some tests fail...

File: include/seqan/index/index_qgram.h (hunk 1 above)
Comment (eseiler) [reply include/seqan/index/index_qgram.h]: I don't really know, but compiling fails if you don't have the non const version...

File: include/seqan/index/index_qgram_openaddressing.h (hunk 2 above)
Comment (eseiler) [reply include/seqan/index/index_qgram_openaddressing.h]: I could also do member init in the ctor,but is was easier to just initialise the member.

File: include/seqan/index/index_qgram.h (hunk 3 above)
Comment (eseiler) [reply include/seqan/index/index_qgram.h]: Same as the other one

File: include/seqan/find/find_myers_ukkonen.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: It seems that we want something like this:

```suggestion
    memset(state.bitMasks, 0, (ValueSize<TValue>::VALUE + 1) * blockCount);
```

(the sizeof seems to be wrong)

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: Oh man I love seqan2:

https://github.com/seqan/seqan/blob/f5f658343c366c9c3d44ba358ffc9317e78a09ed/include/seqan/find/find_myers_ukkonen.h#L216

So the array has actually that size. So it is correct to remove the (+1).

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: So the question is if `|sigma| * w >= (|sigma| + 1) * blockCount =  |sigma| * blockCount + blockCount`

=>

`w >= blockCount + blockCount / |sigma|`

and blockCount is defined as

`block_count = ceiling ( (|query| - 1) / w );`

that means block_count is increasing in query size.

That means a static array is really strange.

File: include/seqan/index/index_qgram_openaddressing.h (hunk 2 above)
Comment (rrahn) [reply include/seqan/index/index_qgram_openaddressing.h]: But the member is initialized in the default c’tor or is it not?

File: include/seqan/find/find_myers_ukkonen.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan/find/find_myers_ukkonen.h]: So it should be replaced at the other places as well?

File: include/seqan/index/index_qgram.h (hunk 1 above)
Comment (rrahn) [reply include/seqan/index/index_qgram.h]: Wow...

File: include/seqan/find/find_myers_ukkonen.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: Okay I love seqan even more:

https://github.com/seqan/seqan/blob/f5f658343c366c9c3d44ba358ffc9317e78a09ed/include/seqan/find/find_myers_ukkonen.h#L360

I remember how bad this module was, the fix as it is seems to be a bit more correct. :)

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: > So it should be replaced at the other places as well?

No, the fix is right as it is.

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan/find/find_myers_ukkonen.h]: :thinking: We might also want to change the static array size to (+1) to match the initial statement.

This might be safer overall.

**Comment by eseiler** [comment]
@rrahn ping
